Add DistributedContext implementation and Delete ITag* Types

tarekgh opened the pull request:
This change includes the following:
- Add the missing implementation of the DistributedContext.
- Change DistributedContextEntry to a struct to avoid allocations inside the DistributedContext.
- Introducing DistributedContext carrier which will carry the current context during the execution
    - By default, DistributedContext will have Noop carrier for optimizing when the SDK is not loaded.
    - When SDK loaded, it should set the carrier to the async local carrier defined in the SDK
-	Removed all Tag* and ITag* types which not needed as DistributedContext should cover the required scenarios.
-	Modified the serialization to work with the DistributedContext.
-	Cleaned up the tests and added more DistributedContext specific tests.

tarekgh commented:
@SergeyKanzhelev I'll try to follow up with the comments on this PR. I just want to say, there is opportunities to optimize more to provide more constructors overload to DistributedContext to have the caller avoid creating any collection. We may look at that later.

tarekgh commented:
I'll look at the failing tests

tarekgh commented:
@lmolkova @SergeyKanzhelev do you think the failing test is related to my changes here? 

https://dev.azure.com/opentelemetry/pipelines/_build/results?buildId=915

```
[xUnit.net 00:00:03.54]     OpenTelemetry.Trace.Export.Test.BatchingSpanProcessorTest.ProcessorDoesNotBlockOnExporter [FAIL]
  X OpenTelemetry.Trace.Export.Test.BatchingSpanProcessorTest.ProcessorDoesNotBlockOnExporter [1s 17ms]
  Error Message:
   Expected at least 1, got 0
Expected: True
Actual:   False
  Stack Trace:
     at OpenTelemetry.Trace.Export.Test.BatchingSpanProcessorTest.WaitForSpans(TestExporter exporter, Int32 spanCount, TimeSpan timeout) in D:\a\1\s\test\OpenTelemetry.Tests\Impl\Trace\Export\BatchingSpanProcessorTests.cs:line 341
   at OpenTelemetry.Trace.Export.Test.BatchingSpanProcessorTest.ProcessorDoesNotBlockOnExporter() in D:\a\1\s\test\OpenTelemetry.Tests\Impl\Trace\Export\BatchingSpanProcessorTests.cs:line 243
Results File: D:\a\_temp\VssAdministrator_fv-az170_2019-12-02_19_14_55.trx
Test Run Failed.
```

lmolkova commented:
@tarekgh I don't think test failures are related to your change. I guess it's still this issue https://github.com/open-telemetry/opentelemetry-dotnet/issues/344. Do you have permission to re-run build?

tarekgh commented:
> Do you have permission to re-run build?

I don't :-(

lmolkova commented:
I've restarted the pipeline. You could also close and reopen PR, it re-triggers pipelines.

SergeyKanzhelev commented on src/OpenTelemetry.Api/DistributedContext/DistributedContext.cs:
```diff
@@ -33,13 +35,113 @@ namespace OpenTelemetry.Context
+        /// </summary>
+        /// <param name="key">The key of the context entry.</param>
+        /// <param name="value">The value of the context entry.</param>
+        public DistributedContext(string key, string value)
```
nice to have this override!

SergeyKanzhelev commented on src/OpenTelemetry/DistributedContext/DistributedContextState.cs:
```diff
@@ -13,17 +13,24 @@
+        internal DistributedContextState(in DistributedContext context)
+        {
+            this.context = AsyncLocalDistributedContextCarrier.Instance.Current;
+            ((AsyncLocalDistributedContextCarrier)AsyncLocalDistributedContextCarrier.Instance).OverwriteCurrent(in context);
```
so values are not inherited? It may be an unexpected behavior for scope API

tarekgh replied:
what you mean by values are not inherited? could you give some example what would be broken by that?

SergeyKanzhelev replied:
if i defined a larger scope with the value "appVersion=1.0" and later smaller scope with the property "flightID=Foo" I want both properties to be propagated downstream. Without manually merging these two collections

tarekgh replied:
Is this always the case? I mean when setting smaller scope, should it merge it with the current larger scope? or this should be be opt-in behavior? I am seeing this is as implementation details but would be easy to support it.

SergeyKanzhelev replied:
scenario wise - it is expected to inherit scopes. But there should be a mechanism to 

1. wipe it and start fresh (typical for situation you enter some "hostile" code that will call into external services)  
2. remove/overwrite certain keys

tarekgh replied:
> wipe it and start fresh (typical for situation you enter some "hostile" code that will call into external services)

This can be done easily by 
``C#
DistributedContext.SetCurrent(DistributedContext.Empty);
```

> remove/overwrite certain keys

Either can be done manually by getting the DistributedContext.Entries and then add/remove/modify. Or we can add some helper methods inside the DistributedContext to do that.

SergeyKanzhelev replied:
https://github.com/open-telemetry/opentelemetry-dotnet/issues/374

tarekgh commented on src/OpenTelemetry.Api/DistributedContext/DistributedContext.cs:
```diff
@@ -33,13 +35,113 @@ namespace OpenTelemetry.Context
         public DistributedContext(IEnumerable<DistributedContextEntry> entries)
         {
-            this.entries = entries;
+            if (entries is null)
```
ok, I'll do that.

tarekgh replied:
This is done!

tarekgh commented on src/OpenTelemetry.Api/DistributedContext/DistributedContext.cs:
```diff
@@ -33,13 +35,113 @@ namespace OpenTelemetry.Context
+            {
+                if (entry == default)
+                {
+                    throw new ArgumentException($"'{nameof(entries)}' contains entries with null key and value");
```
ok, I'll do that.

tarekgh replied:
Done!

tarekgh commented on src/OpenTelemetry.Api/DistributedContext/DistributedContext.cs:
```diff
@@ -33,13 +35,113 @@ namespace OpenTelemetry.Context
+                    List<DistributedContextEntry> list = new List<DistributedContextEntry>();
+
+                    // Remove the duplicates keys.
+                    foreach (DistributedContextEntry entry in entries)
```
I did that because there were some tests were validating this behavior.

tarekgh replied:
After more thoughts I decided to keep the filtering for now for the sake of not complicating the equality and avoid doing filtering every time we serialize the same context.

tarekgh commented on src/OpenTelemetry.Api/DistributedContext/DistributedContext.cs:
```diff
@@ -33,13 +35,113 @@ namespace OpenTelemetry.Context
+                    List<DistributedContextEntry> list = new List<DistributedContextEntry>();
+
+                    // Remove the duplicates keys.
+                    foreach (DistributedContextEntry entry in entries)
```
After I remove the null check we'll be enumerating once. I'll fix that.

tarekgh replied:
After more thoughts I decided to keep the filtering for now for the sake of not complicating the equality and avoid doing filtering every time we serialize the same context.

tarekgh commented on src/OpenTelemetry.Api/DistributedContext/DistributedContext.cs:
```diff
@@ -33,13 +35,113 @@ namespace OpenTelemetry.Context
+            get => carrier;
+            set
+            {
+                if (value is null)
```
Are you suggesting to have like DistributedContextFactory which create DistributedContext objects and allow setting the carrier there? if so, this will require creating many other abstracting classes and really will complicate it. I am treating DistributedContext as special because we defined it not abstracted at all and self contained.  I am not sure why the factory pattern is better here. please clarify if I am missing any point.

SergeyKanzhelev replied:
https://github.com/open-telemetry/opentelemetry-dotnet/issues/373

tarekgh commented:
@SergeyKanzhelev I have addressed some of your feedback. 

Regarding filtering the duplicate entries in the DistributedContext construction, I have kept the filtering but optimized more. The reason is not filtering will add more complexity in other functionality like Equality. Also, if we didn't filter it we'll have to filter every time the context get serialized which not good to do that for the same context every time you serialize it. Let's keep the filtering for now and we can optimize more later if needed.
I'll wait your answers for the other opened comments you raised.

Thanks for the feedback.

SergeyKanzhelev commented:
@tarekgh I appreciate you working on your vacation. I tend to accept this PR and follow up on changes later. @open-telemetry/dotnet-approvers any objections?

MikeGoldsmith commented:
I think this is good to accept.
